src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -408,5 +381,46 @@ public ValueTask<(Solution solution, bool updated)> TryUpdateWorkspaceCurrentSol
                 return tuple.solution;
             }
         }
+
+        private sealed class LazySolution : IAsyncDisposable, IDisposable
+        {
+            private readonly RemoteWorkspace _remoteWorkspace;
+            private readonly Checksum _solutionChecksum;
+            private readonly CancellationTokenSource _cancellationTokenSource = new();
+            private readonly Task<Solution> _task;
+
+            public LazySolution(RemoteWorkspace remoteWorkspace, Checksum solutionChecksum, Func<CancellationToken, Task<Solution>> getSolutionAsync)
+            {
+                _remoteWorkspace = remoteWorkspace;
+                _solutionChecksum = solutionChecksum;
+                _task = getSolutionAsync(_cancellationTokenSource.Token);
+            }
+
+            public Task<Solution> Task => _task;
+
+            void IDisposable.Dispose()
+            {
+                // This type is only used as IAsyncDisposable, but needs to implement IDisposable for wrapping within
+                // ReferenceCountedDisposable<T>.
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            public async ValueTask DisposeAsync()
+            {
+                _cancellationTokenSource.Cancel();
+                _cancellationTokenSource.Dispose();
+
+                // We use CancellationToken.None here as we have to ensure the lazy solution is removed from the
+                // checksum map, or else we will have a memory leak.  This should hopefully not ever be an issue as we
+                // only ever hold this gate for very short periods of time in order to set do basic operations on our
+                // state.
+                using (await _remoteWorkspace._gate.DisposableWaitAsync(CancellationToken.None).ConfigureAwait(false))
+                {
+                    _remoteWorkspace._solutionChecksumToLazySolution.Remove(_solutionChecksum);
+                }
+
+                await _task.NoThrowAwaitable(false);
```

Comment:
ah, i think i got it.  this is the critical line it seems.  the last one who actually does the dispose will wait on the task being complete.   i think this needs doc.s

Reply:
plz doc why NoThrowAwaitable.